### PR TITLE
Instrument obsreport.receiver metrics with otel go

### PR DESCRIPTION
**Pull request by paivagustavo**
**Description:** Instruments existing `obsreport.Receiver` metrics  with otel-go, side-by-side with the existing opencensus metrics.

This uses the same `UseOtelForInternalMetricsfeatureGateID` feature gate, adding a `UseOtelForMetrics()` method on the `obsreportconfig` that should be used by components of the collector core to check what sdk should be used for instrumentation.

**Link to tracking Issue:** part of https://github.com/open-telemetry/opentelemetry-collector/issues/816

**Testing:** Manual tests comparing the current exported metrics and the metrics from the otel-go sdk after enabling the feature gate `--feature-gates=telemetry.useOtelForInternalMetrics`

### CODE REVIEW

File: internal/obsreportconfig/obsmetrics/obs_receiver.go [hunk 1]
```diff
@@ -83,4 +87,40 @@ var (
 		ReceiverPrefix+RefusedLogRecordsKey,
 		"Number of log records that could not be pushed into the pipeline.",
 		stats.UnitDimensionless)
+
+	otelMeter     = global.Meter(MeterName)
+	AttrReceiver  = attribute.Key(ReceiverKey)
+	AttrTransport = attribute.Key(TransportKey)
```
Comment (bogdandrutu) [inline internal/obsreportconfig/obsmetrics/obs_receiver.go:93]: Why using the global meter?

File: internal/obsreportconfig/obsmetrics/obs_receiver.go [hunk 2]
```diff
@@ -83,4 +87,40 @@ var (
...
+	OtelReceiverAcceptedLogRecords, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+AcceptedLogRecordsKey,
+		instrument.WithDescription("Number of log records successfully pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
+	OtelReceiverRefusedLogRecords, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+RefusedLogRecordsKey,
+		instrument.WithDescription("Number of log records that could not be pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
```
Comment (bogdandrutu) [inline internal/obsreportconfig/obsmetrics/obs_receiver.go:125]: these can be defined in the obsreport package now, since we don't need to share them between other internal packages.

File: obsreport/obsreport_receiver.go [hunk 3]
```diff
@@ -163,23 +175,42 @@ func (rec *Receiver) endOp(
...
+			switch dataType {
+			case config.TracesDataType:
+				acceptedMeasure = obsmetrics.OtelReceiverAcceptedSpans
+				refusedMeasure = obsmetrics.OtelReceiverRefusedSpans
+			case config.MetricsDataType:
+				acceptedMeasure = obsmetrics.OtelReceiverAcceptedMetricPoints
+				refusedMeasure = obsmetrics.OtelReceiverRefusedMetricPoints
+			case config.LogsDataType:
+				acceptedMeasure = obsmetrics.OtelReceiverAcceptedLogRecords
+				refusedMeasure = obsmetrics.OtelReceiverRefusedLogRecords
```
Comment (bogdandrutu) [inline obsreport/obsreport_receiver.go:189]: extract separate funcs "recordWithOtel" and "recordWithOC" to help with readability.

File: obsreport/views.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+var storage = views{
+	v: []otelview.View{},
+}
+
+// AddViews to the OpenTelemetry Go SDK.
+// Views are only configurable at the SDK initialization, for this reason
+// components must call this function inside an init() function to be correctly initialized.
+func AddViews(v ...otelview.View) {
```
Comment (bogdandrutu) [inline obsreport/views.go:47]: Why do you want to expose this functionality? What is the purpose of it?

**Review by bogdandrutu — COMMENTED** [review]
We need this https://github.com/open-telemetry/opentelemetry-collector/pull/5939, which is to not break the metrics names at this time. We will at one point later revisit all names.

File: obsreport/views.go (hunk 4 above)
Comment (bogdandrutu) [reply obsreport/views.go:47]: This can/should be added later. There are other nicer alternatives like allowing Factories to implement additional interfaces for this, or doing it in the NewFactory helper as optionals, etc.

Comment (paivagustavo) [reply obsreport/views.go]: I'll just remove this for now as this was just here as a preparation step migrations of some other components, thanks for sharing some nicer options as well!

File: obsreport/obsreport_receiver.go (hunk 3 above)
Comment (paivagustavo) [reply obsreport/obsreport_receiver.go]: done.

File: internal/obsreportconfig/obsmetrics/obs_receiver.go (hunk 1 above)
Comment (paivagustavo) [reply internal/obsreportconfig/obsmetrics/obs_receiver.go]: I've moved those to be created inside of `obsreport.Receiver` using the `ReceiverCreateSettings.MeterProvider`.

File: obsreport/obsreport_receiver.go [hunk 5]
```diff
@@ -63,7 +87,70 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
...
+
+func (rec *Receiver) createOtelMetrics() {
+	var err error
+	handleError := func(metricName string, err error) {
+		if err != nil {
+			rec.logger.Warn("failed to create otel instrument", zap.Error(err), zap.String("metric", metricName))
+		}
 	}
+
+	rec.acceptedSpansCounter, err = rec.meter.SyncInt64().Counter(
```
Comment (jaronoff97) [inline obsreport/obsreport_receiver.go:113]: thought: I wonder if these should exist in a list/map that are then just accessed as needed. i.e.

```
        telemetryMap = map[config.DataType]map[string]*syncint64.Counter
        // iterate through [DataType] x ["accepted", "refused"] to initialize
...
	telemetryMap[dataType]["accepted"].Add(receiverCtx, int64(numAccepted), rec.otelAttrs...)
	telemetryMap[dataType]["refused"].Add(receiverCtx, int64(numRefused), rec.otelAttrs...)
```

**Review by jaronoff97 — APPROVED** [review]
Just one thought, otherwise looks great!

**Comment by paivagustavo** [comment]
> We need this https://github.com/open-telemetry/opentelemetry-collector/pull/5939, which is to not break the metrics names at this time. We will at one point later revisit all names.

@bogdandrutu  this pr actually fixes the name inconsistencies wrapping the `prometheus.Registry`, this might be easier than trying add `otelcol_` prefix to all metrics (including contrib metrics):
```
prometheus.WrapRegistererWithPrefix("otelcol_", prometheus.WrapRegistererWith(promLabels, registry))
```

If we still wants to move on with #5939, I can create a copy of `obsmetrics.ReceiverPrefix`, so this won't make `otelcol_otelcol_` metrics and fix this after that get merged.

File: internal/obsreportconfig/obsmetrics/obsmetrics.go [hunk 6]
```diff
@@ -19,4 +19,10 @@ package obsmetrics // import "go.opentelemetry.io/collector/internal/obsreportco
 
 const (
 	NameSep = "/"
+
+	MeterName = "io.opentelemetry.collector"
```
Comment (bogdandrutu) [inline internal/obsreportconfig/obsmetrics/obsmetrics.go:23]: Why this name? Do we have the same name for receivers/exporters? How does this affect the metrics exposed to Prometheus? /cc @dashpole 

In order to be a no-op change between OC -> Otel for the moment, I think we should have an empty name for the moment. Then when we have the final list of the metrics, we should agree on what scope name we use.

File: obsreport/obsreport.go [hunk 7]
```diff
@@ -17,10 +17,17 @@ package obsreport // import "go.opentelemetry.io/collector/obsreport"
...
+	"go.opentelemetry.io/collector/internal/obsreportconfig/obsmetrics"
 )
 
 func recordError(span trace.Span, err error) {
 	if err != nil {
 		span.SetStatus(codes.Error, err.Error())
 	}
 }
+
+func InstrumentWithOtel() bool {
```
Comment (bogdandrutu) [inline obsreport/obsreport.go:31]: Why public? Who is the user of this?

File: obsreport/obsreport_receiver.go (hunk 5 above)
Comment (bogdandrutu) [reply obsreport/obsreport_receiver.go:113]: Accessing a map is slower than direct access. So I prefer this way.

File: service/telemetry.go [hunk 8]
```diff
@@ -213,15 +221,39 @@ func (tel *telemetryInitializer) initOpenCensus(cfg telemetry.Config, telAttrs m
...
 	registry := prometheus.NewRegistry()
-	if err := registry.Register(exporter.Collector); err != nil {
+
+	promLabels := make(prometheus.Labels)
+	for k, v := range attrs {
+		promLabels[sanitizePrometheusKey(k)] = v
+	}
+
+	wrappedRegisterer := prometheus.WrapRegistererWithPrefix("otelcol_", prometheus.WrapRegistererWith(promLabels, registry))
+	if err := wrappedRegisterer.Register(exporter.Collector); err != nil {
```
Comment (bogdandrutu) [inline service/telemetry.go:256]: Let's split this change in a separate PR.

Comment (paivagustavo) [reply service/telemetry.go:256]: Created https://github.com/open-telemetry/opentelemetry-collector/pull/6223 for this.

File: internal/obsreportconfig/obsmetrics/obsmetrics.go (hunk 6 above)
Comment (paivagustavo) [reply internal/obsreportconfig/obsmetrics/obsmetrics.go:23]: This can be removed since I'm no longer using the global meter and using the `ReceiverCreateSettings.MeterProvider` instead, there I'm using the same name as the trace name which is `cfg.ReceiverID.String()`.


> In order to be a no-op change between OC -> Otel for the moment, I think we should have an empty name for the moment.

I think this does not matter for the prometheus exporter,  as the otel scope is ignored now and will be added as additional attributes and metrics per https://github.com/open-telemetry/opentelemetry-specification/pull/2703

File: obsreport/obsreport.go (hunk 7 above)
Comment (paivagustavo) [reply obsreport/obsreport.go:31]: I planned this to be public for all components that want to double instrument with otel-go, but as I've discovered the  `TelemetrySettings` struct, that seems to be a more suitable place, what do you think about that?

**Review by jpkrohling — COMMENTED** [review]
This looks sane to me,  a review from @Aneurysm9 would be helpful though.

File: component/telemetry.go [hunk 9]
```diff
@@ -33,6 +33,10 @@ type TelemetrySettings struct {
 	// MeterProvider that the factory can pass to other instrumented third-party libraries.
 	MeterProvider metric.MeterProvider
 
+	// UseOtelForMetrics controls whether the collector should use otel go sdk for
+	// internal telemetry metrics.
+	UseOtelForMetrics bool
```
Comment (bogdandrutu) [inline component/telemetry.go:38]: Why exposing this?

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Current PR exposes new temporary public APIs because the OC bridge is not ready and we cannot really start the proper migration to otel. 

I am not OK doing that.

**Comment by paivagustavo** [comment]
@bogdandrutu  are you concerned about having this public `TelemetrySettings.UseOtelForMetrics`? Would it be better if we get rid of that and just started to record values to both sdks at the same time? Something like this:
```go
rec.recordWithOtel(receiverCtx, dataType, numAccepted, numRefused)
rec.recordWithOC(receiverCtx, dataType, numAccepted, numRefused)
```

I'm trying to understand what other paths we might have other than just waiting for the bridge to get ready.

**Comment by bogdandrutu** [comment]
@paivagustavo to give you alternatives, I need to understand the goal :)

File: component/telemetry.go (hunk 9 above)
Comment (paivagustavo) [reply component/telemetry.go]: I've moved it to the internal package `obsreportconfig` so we can use it to the collector core component with otel.

File: internal/obsreportconfig/obsreportconfig.go [hunk 10]
```diff
@@ -20,9 +20,21 @@ import (
...
 	"go.opentelemetry.io/collector/internal/obsreportconfig/obsmetrics"
 )
 
+const (
+	// UseOtelForInternalMetricsfeatureGateID is the feature gate ID that controls whether the collector uses open
+	// telemetrySettings for internal metrics.
+	UseOtelForInternalMetricsfeatureGateID = "telemetry.useOtelForInternalMetrics"
+)
+
+// UseOtelForMetrics controls whether the Collector should use Otel for Collector's internal instrumentation.
```
Comment (codeboten) [inline internal/obsreportconfig/obsreportconfig.go:33]: nit: this method doesn't control anything but reports it 

```suggestion
// UseOtelForMetrics returns whether the Collector should use Otel for Collector's internal instrumentation.
```

File: obsreport/obsreport_receiver.go [hunk 11]
```diff
@@ -36,6 +42,19 @@ type Receiver struct {
 	longLivedCtx   bool
 	mutators       []tag.Mutator
 	tracer         trace.Tracer
+	meter          metric.Meter
+	logger         *zap.Logger
+
+	// Otel API
```
Comment (codeboten) [inline obsreport/obsreport_receiver.go:48]: Is this comment necessary here?

File: obsreport/obsreport_receiver.go [hunk 12]
```diff
@@ -63,7 +83,70 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
...
+		logger: cfg.ReceiverCreateSettings.Logger,
+
+		useOtelForMetrics: obsreportconfig.UseOtelForMetrics(),
+		otelAttrs: []attribute.KeyValue{
+			attribute.String(obsmetrics.ReceiverKey, cfg.ReceiverID.String()),
+			attribute.String(obsmetrics.TransportKey, cfg.Transport),
+		},
 	}
+
+	rec.createOtelMetrics()
```
Comment (codeboten) [inline obsreport/obsreport_receiver.go:96]: should this be behind a check for using otel metrics?

**Review by codeboten — COMMENTED** [review]
Just a couple of questions

File: service/telemetry.go [hunk 13]
```diff
@@ -85,7 +81,7 @@ func init() {
 // registerInternalMetricFeatureGate registers the Internal Metric feature gate to the passed in registry
 func registerInternalMetricFeatureGate(registry *featuregate.Registry) {
 	registry.MustRegister(featuregate.Gate{
-		ID:          useOtelForInternalMetricsfeatureGateID,
+		ID:          obsreportconfig.UseOtelForInternalMetricsfeatureGateID,
```
Comment (Aneurysm9) [inline service/telemetry.go:84]: This registration should move to `obsreportconfig` if the ID constant is moving there as well.

File: internal/obsreportconfig/obsreportconfig.go (hunk 10 above)
Comment (Aneurysm9) [reply internal/obsreportconfig/obsreportconfig.go:33]: I don't think this function is even needed since the feature gate ID constant is also exported.  It's easy enough for consumers to check the gate themselves and also hopefully avoids hiding the fact that a gate is being checked so it isn't called in a tight loop accidentally.

File: obsreport/obsreport_receiver.go (hunk 12 above)
Comment (Aneurysm9) [reply obsreport/obsreport_receiver.go:96]: Or the check should be at the top of the method.  One of the two.

File: obsreport/obsreport_receiver.go [hunk 14]
```diff
@@ -163,23 +246,11 @@ func (rec *Receiver) endOp(
...
-		case config.TracesDataType:
-			acceptedMeasure = obsmetrics.ReceiverAcceptedSpans
-			refusedMeasure = obsmetrics.ReceiverRefusedSpans
-		case config.MetricsDataType:
-			acceptedMeasure = obsmetrics.ReceiverAcceptedMetricPoints
-			refusedMeasure = obsmetrics.ReceiverRefusedMetricPoints
-		case config.LogsDataType:
-			acceptedMeasure = obsmetrics.ReceiverAcceptedLogRecords
-			refusedMeasure = obsmetrics.ReceiverRefusedLogRecords
+		if rec.useOtelForMetrics {
```
Comment (Aneurysm9) [inline obsreport/obsreport_receiver.go:249]: Maybe hide this behind a `rec.record()` method that does the switching.

File: obsreport/obsreport_receiver.go (hunk 12 above)
Comment (paivagustavo) [reply obsreport/obsreport_receiver.go:96]: I've decided to add a check at the top.

File: obsreport/obsreport_receiver.go (hunk 14 above)
Comment (paivagustavo) [reply obsreport/obsreport_receiver.go]: Created a `recordMetrics` to hide it and also to not be mistaken with the `if span.IsRecording() ` line below which is not refer a different type of "record"

File: service/telemetry.go (hunk 13 above)
Comment (paivagustavo) [reply service/telemetry.go]: done.

File: obsreport/obsreport_receiver.go (hunk 11 above)
Comment (paivagustavo) [reply obsreport/obsreport_receiver.go]: Not really 🙃

File: internal/obsreportconfig/obsreportconfig.go (hunk 10 above)
Comment (paivagustavo) [reply internal/obsreportconfig/obsreportconfig.go]: Just removed it.

**Comment by paivagustavo** [comment]
Thanks for the reviews @bogdandrutu @codeboten @Aneurysm9, I think I've addressed all the comments at this point and it should be ready to be reviewed again.

**Review by codeboten — APPROVED** [review]
Thanks for addressing my comments 👍🏻

**Review by Aneurysm9 — APPROVED** [review]

File: internal/obsreportconfig/obsreportconfig.go [hunk 15]
```diff
@@ -20,9 +20,30 @@ import (
...
+func init() {
+	// register feature gate
+	RegisterInternalMetricFeatureGate(featuregate.GetRegistry())
+}
+
+// RegisterInternalMetricFeatureGate registers the Internal Metric feature gate to the passed in registry
+func RegisterInternalMetricFeatureGate(registry *featuregate.Registry) {
+	registry.MustRegister(featuregate.Gate{
+		ID:          UseOtelForInternalMetricsfeatureGateID,
+		Description: "controls whether the collector to uses OpenTelemetry for internal metrics",
```
Comment (evantorrie) [inline internal/obsreportconfig/obsreportconfig.go:42]:
```suggestion
		Description: "controls whether the collector uses OpenTelemetry for internal metrics",
```

File: obsreport/obsreport_receiver.go [hunk 16]
```diff
@@ -63,7 +83,74 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
 			tag.Upsert(obsmetrics.TagKeyTransport, cfg.Transport, tag.WithTTL(tag.TTLNoPropagation)),
 		},
 		tracer: cfg.ReceiverCreateSettings.TracerProvider.Tracer(cfg.ReceiverID.String()),
+		meter:  cfg.ReceiverCreateSettings.MeterProvider.Meter(cfg.ReceiverID.String()),
```
Comment (bogdandrutu) [inline obsreport/obsreport_receiver.go:86]: Not sure if this is the right name for the meter. Should we have a meter per factory type of per factory component instance?

Comment (paivagustavo) [reply obsreport/obsreport_receiver.go]: Do you mean to use `receiver/otlp` for the `otlp` receiver's meter name instead of just `otlp`, to differentiate components that have the same name? 

If that is what you meant, I think that seems better. Should tracer name be like that as well?

File: internal/obsreportconfig/obsreportconfig.go [hunk 17]
```diff
@@ -103,6 +116,26 @@ func allViews() []*view.View {
 	return views
 }
 
+func receiverViews() []*view.View {
+	if featuregate.GetRegistry().IsEnabled(UseOtelForInternalMetricsfeatureGateID) {
```
Comment (paivagustavo) [inline internal/obsreportconfig/obsreportconfig.go:120]: kinda misc, but I've added featuregate check on the OC receiver views creation.

File: obsreport/obsreport_receiver.go (hunk 16 above)
Comment (bogdandrutu) [reply obsreport/obsreport_receiver.go]: Nope, I mean to use `cfg.ReceiverID.Type()`, so if we have multiple instances of the `otlp` receiver we have same scope for this, and the instance  name is not in the scope name (probably should by an attribute).

Comment (Aneurysm9) [reply obsreport/obsreport_receiver.go]: I don't think `otlp` is an appropriate scope name.  Shouldn't the scope here be `go.opentelemetry.io/collector/obsreport` with attributes for `componentType=receiver` and `componentID=otlp` (or `componentID=otlp/foo` if a name exists on the receiver ID)?

Comment (bogdandrutu) [reply obsreport/obsreport_receiver.go]: That is also a possibility, but you are still suggesting the same thing, that component name should not be part of the scope (which was before).

I think I like scope.name=`go.opentelemetry.io/collector/obsreport/receiver` and have a `component.id=otlp` (or `component.id=otlp/foo`) since we use `.` mostly as attribute name separator in otel (see resource attributes for example).

File: internal/obsreportconfig/obsreportconfig.go (hunk 17 above)
Comment (bogdandrutu) [reply internal/obsreportconfig/obsreportconfig.go:120]: why? I don't think we should code this way, the contract is to register these views only if `UseOtelForInternalMetricsfeatureGateID`, then we protect the high level function we don't have to protect any sub-function.

Comment (paivagustavo) [reply internal/obsreportconfig/obsreportconfig.go:120]: Because if we want to use `prometheus.Registry` as the bridge, we don't want OC to emit metrics that have being migrated to otel-go if the featuregate is enabled, but we would want all the others non-migrated oc metrics to be exported as normal, hence I just removed the `receiver` metrics for now.

Once all components of the core gets instrumented with otel-go, we can move this to be broader and not register any view at all when the featuregate is enabled, but for now I would suggest to just remove it component per component.

File: obsreport/obsreport_receiver.go (hunk 16 above)
Comment (paivagustavo) [reply obsreport/obsreport_receiver.go]: By the way, we will encounter some warns/missing metrics due to some otel-go bugs at the moment. 

Because the go sdk doesn't return a valid instrument for "duplicate" metric creation results in invalid instruments in otel-go ([tracking issue](https://github.com/open-telemetry/opentelemetry-go/issues/3229
)), only the first receiver would have valid instruments and subsequent calls to the `obsreport.NewReceiver()` will have noop metrics.

I would suggest us to just add the `component.id` (`cfg.ReceiverID.String()`) as suffix for now, adding a TODO with the tracking issue and remove that once that gets fixed. Since the prometheus exporter doesn't use the scope for nothing yet, it wouldn't change anything.

File: obsreport/obsreport_receiver.go [hunk 18]
```diff
@@ -63,7 +88,74 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
...
+func (rec *Receiver) createOtelMetrics() {
+	if !rec.useOtelForMetrics {
+		return
+	}
+
+	var err error
+	handleError := func(metricName string, err error) {
+		if err != nil {
+			rec.logger.Warn("failed to create otel instrument", zap.Error(err), zap.String("metric", metricName))
+		}
```
Comment (bogdandrutu) [inline obsreport/obsreport_receiver.go:115]: We need to return the error to the caller instead of logging. It should not be done in this PR, but we need to file an issue for this.

**Review by bogdandrutu — APPROVED** [review]

File: obsreport/obsreport_receiver.go (hunk 16 above)
Comment (bogdandrutu) [reply obsreport/obsreport_receiver.go]: I would not do that, since that will be a breaking change. But fix was release today.

Comment (paivagustavo) [reply obsreport/obsreport_receiver.go]: Right, upgrading to the new released version fixed this.

**Review by bogdandrutu — APPROVED** [review]
Just for my sanity, please create a separate PR for the version updates, maybe also include the 1.11 updates as well.
